### PR TITLE
Add read_into_dict function for simpler CSVY parsing

### LINKED ISSUE #91: Include a native Python dict `read` function
Since the most basic version of this package only requires the pyyaml dependency, there should be a basic `read` function that returns a native Python type.

There is currently a `read_to_list` function (that isn't exposed as a top-level function), that returns a nested list. But that isn't in the most developer friendly format because it has a mix of column headers and data in the list. A dictionary type feels more appropriate.

The dictionary could be structured with 2 keys, "columns" and "data", with "data" being a 2D nested list. Or it could have the columns names as keys and the values each being a list for that column.

### PR DESCRIPTION
**Pull request by Kaos599**
Closes #91 

- **Refactored `read_into_dict` Function:**
  - Implemented a structured approach to read CSVY files, extracting both metadata and data sections.
  - Added error handling for cases where no metadata is found, returning an empty dictionary and printing a warning instead of raising an error.
  - Updated the function's docstring to clarify assumptions about the data structure, specifying that the first row contains column names.

- **Updated Tests for `read_into_dict`:**
   - Converted tests to dynamically handle files with or without metadata, outputting a `UserWarning` for files lacking metadata rather than raising assertions.
   - Implemented `pytest.mark.parametrize` to test multiple datasets seamlessly, enabling flexibility and scalability.
   - Updated metadata checks to verify general presence instead of specific keys, ensuring broader file compatibility.
   - Added logging to debug metadata contents and data types, aiding developers in tracing issues effectively.

- **Improved Logging:**
  - Integrated logging throughout the `read_into_dict` function and the test suite to track processing steps and outcomes, aiding in future debugging efforts.

- **Code Formatting and Compliance:**
  - Addressed code style issues, including import organization and compliance with PEP 8 guidelines.
  - Ensured that all functions and test cases have appropriate docstrings and are properly formatted.

### CODE REVIEW

**Comment by Kaos599** [comment]
@dalonsoa @AdrianDAlessandro 

I think everything is fine except 
this error by Codecov.
``` 
csvy/readers.py#L307-L308
Added lines #L307 - L308 were not covered by tests
```

Refering to these [lines](https://github.com/Kaos599/pycsvy/blob/7ea22bab51d9b42fef816ff78af7ee1a651df1c8/csvy/readers.py#L307-L308).

File: csvy/readers.py [hunk 1]
```diff
@@ -270,3 +271,52 @@ def read_to_list(
...
+                break
+            continue
+
+        if section == "metadata":
+            if ":" in line:
+                key, value = line.split(":", 1)
+                metadata[key.strip()] = value.strip()
+        elif section == "data":
+            if line:
+                data.append(line.split(","))
```
Comment (AdrianDAlessandro) [inline csvy/readers.py:316]: This section is essentially reimplementing what we already have in the other functions in this module. They should be used to ensure consistent behaviour

File: csvy/readers.py [hunk 2]
```diff
@@ -270,3 +271,52 @@ def read_to_list(
             data.append(row)
 
     return data, header
+
+
+def read_into_dict(file_path: Path | str):
```
Comment (AdrianDAlessandro) [inline csvy/readers.py:276]: This should have the same input parameters as the other read functions to maintain a consistent API.

File: tests/data_read_into_dict.csvy
Comment (AdrianDAlessandro) [inline tests/data_read_into_dict.csvy:1]: This file isn't necessary, we already have test  data files.

File: tests/test_read_into_dict.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+data_files = [
+    "tests/data_comment.csv",
+    "tests/array_data.csv",
+    "tests/data.csv",
+    "tests/data_read_into_dict.csvy",
+]
+
+
+@pytest.mark.parametrize("file_path", data_files)
+def test_read_into_dict(file_path):
```
Comment (AdrianDAlessandro) [inline tests/test_read_into_dict.py:22]: This test function should be moved to `test_read.py`, which is where all the test files for the readers are.

Also, it should use the existing fixtures for the data file paths. The `data_path` fixture alone is probably satisfactory for testing the function.

File: tests/test_read_into_dict.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    """Test the read_into_dict function with multiple datasets."""
+    data, metadata = read_into_dict(file_path)
```
Comment (AdrianDAlessandro) [inline tests/test_read_into_dict.py:24]: These values are never checked properly. The content of the `data` and `metadata` variables should be checked in a similar way to how they are checked in the other tests in `test_read.py`

File: csvy/readers.py [hunk 5]
```diff
@@ -270,3 +271,52 @@ def read_to_list(
...
+                key, value = line.split(":", 1)
+                metadata[key.strip()] = value.strip()
+        elif section == "data":
+            if line:
+                data.append(line.split(","))
+
+    if data:
+        data = list(map(list, zip_longest(*data, fillvalue="")))
+        data_dict = {data[i][0]: data[i][1:] for i in range(len(data))}
+        return data_dict, metadata
```
Comment (AdrianDAlessandro) [inline csvy/readers.py:321]: I believe this section of code should work, but I haven't checked it yet because of the other issues.

**Comment by dalonsoa** [comment]
Hi @Kaos599 . As you have not addressed the changes requested by @AdrianDAlessandro , I'm closing this PR without merging. Maybe another time.

**Pull request closed without merging**